src/inmanta/ast/statements/call.py

```diff
@@ -277,11 +281,24 @@ def __init__(self, queue_scheduler, resolver, result: ResultVariable, requires,
         for r in requires.values():
             self.waitfor(r)
         self.ready(self)
+        # resolved args and kwargs
+        self.args = None
+        self.kwargs = None
 
     def execute(self) -> None:
-        requires = {k: v.get_value() for (k, v) in self.base_requires.items()}
+        # Execution in two stages to prevent re-execution of argument expressions
+        if self.args is None:
+            # stage one: arguments
+            # execute once and cache results
+            requires = {k: v.get_value() for (k, v) in self.base_requires.items()}
+            try:
+                self.args, self.kwargs = self.function.execute_args(requires, self.resolver, self.queue)
+            except RuntimeException as e:
+                e.set_statement(self.function)
+                raise e
+
         try:
-            self.function.resume(requires, self.resolver, self.queue, self.result)
+            self.function.execute_call(self.args, self.kwargs, self.resolver, self.queue, self.result)
```

Comment:
Are we confident that the `FunctionUnit` object gets garbage collected or should we set `self.args` and `self.kwargs` back to `None`?